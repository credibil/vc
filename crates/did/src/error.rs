/// DID resolution error codes
#[derive(thiserror::Error, Debug)]
pub enum Error {
    /// The DID method is not supported.
    #[error("methodNotSupported")]
    MethodNotSupported(String),

    /// The DID supplied to the DID resolution function does not conform to
    /// valid syntax.
    #[error("invalidDid")]
    InvalidDid(String),

    /// The DID resolver was unable to find the DID document resulting from
    /// this resolution request.
    #[error("notFound")]
    NotFound(String),

    /// The representation requested via the accept input metadata property is
    /// not supported by the DID method and/or DID resolver.
    #[error("representationNotSupported")]
    RepresentationNotSupported(String),

    /// The DID URL is invalid
    #[error("invalidDidUrl")]
    InvalidDidUrl(String),

    // ---- Creation Errors ----  //
    /// The byte length of raw public key does not match that expected for the
    /// associated multicodecValue.
    #[error("invalidPublicKeyLength")]
    InvalidPublicKeyLength(String),

    /// The public key is invalid
    #[error("invalidPublicKey")]
    InvalidPublicKey(String),

    /// Public key format is not known to the implementation.
    #[error("unsupportedPublicKeyType")]
    UnsupportedPublicKeyType(String),

    /// Other, unspecified errors.
    #[error(transparent)]
    Other(#[from] anyhow::Error),
}

impl Error {
    /// Returns the error code.
    #[must_use]
    pub fn code(&self) -> String {
        self.to_string()
    }

    /// Returns the associated error message.
    #[must_use]
    pub fn message(&self) -> String {
        match self {
            Self::MethodNotSupported(msg)
            | Self::InvalidDid(msg)
            | Self::NotFound(msg)
            | Self::InvalidDidUrl(msg)
            | Self::RepresentationNotSupported(msg)
            | Self::InvalidPublicKeyLength(msg)
            | Self::InvalidPublicKey(msg)
            | Self::UnsupportedPublicKeyType(msg) => msg.clone(),
            Self::Other(err) => err.to_string(),
        }
    }
}

// impl From<anyhow::Error> for Error {
//     fn from(err: anyhow::Error) -> Self {
//         Self
//     }
// }

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn error_code() {
        let err = Error::MethodNotSupported("Method not supported".into());
        assert_eq!(err.message(), "Method not supported");
    }
}
