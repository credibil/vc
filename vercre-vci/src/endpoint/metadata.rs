//! # Metadata Endpoint
//!
//! The Credential Issuer Metadata contains information on the Credential Issuer's
//! technical capabilities, supported Credentials, and (internationalized) display
//! information.
//!
//! The Credential Issuer's configuration can be retrieved using the Credential Issuer
//! Identifier.
//!
//! Credential Issuers publishing metadata MUST make a JSON document available at the
//! path formed by concatenating the string `/.well-known/openid-credential-issuer` to
//! the Credential Issuer Identifier. If the Credential Issuer value contains a path
//! component, any terminating / MUST be removed before appending
//! `/.well-known/openid-credential-issuer`.
//!
//! The language(s) in HTTP Accept-Language and Content-Language Headers MUST use the values defined in [RFC3066](https://www.rfc-editor.org/rfc/rfc3066).
//!
//! Below is a non-normative example of a Credential Issuer Metadata request:
//!
//! ```http
//! GET /.well-known/openid-credential-issuer HTTP/1.1
//!     Host: server.example.com
//!     Accept-Language: fr-ch, fr;q=0.9, en;q=0.8, de;q=0.7, *;q=0.5
//! ```

use tracing::{instrument, trace};
use vercre_core::vci::{MetadataRequest, MetadataResponse};
use vercre_core::{Callback, Client, Holder, Issuer, Result, Server, Signer, StateManager};

use super::Endpoint;

impl<P> Endpoint<P>
where
    P: Client + Issuer + Server + Holder + StateManager + Signer + Callback + Clone,
{
    /// Metadata request handler.
    ///
    /// # Errors
    ///
    /// Returns an `OpenID4VP` error if the request is invalid or if the provider is
    /// not available.
    pub async fn metadata(&self, request: impl Into<MetadataRequest>) -> Result<MetadataResponse> {
        let request = request.into();
        self.handle_request(request, Context {}).await
    }
}

#[derive(Debug)]
struct Context;

impl super::Context for Context {
    type Request = MetadataRequest;
    type Response = MetadataResponse;

    // TODO: get callback_id from state
    fn callback_id(&self) -> Option<String> {
        None
    }

    #[instrument]
    async fn process<P>(&self, provider: &P, request: &Self::Request) -> Result<Self::Response>
    where
        P: Client + Issuer + Server + Holder + StateManager + Signer + Callback + Clone,
    {
        trace!("Context::process");

        // TODO: add languages to request
        let credential_issuer = Issuer::metadata(provider, &request.credential_issuer).await?;
        Ok(MetadataResponse { credential_issuer })
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot as assert_snapshot;
    use test_utils::vci_provider::{Provider, ISSUER};

    use super::*;

    #[tokio::test]
    async fn metadata_ok() {
        test_utils::init_tracer();

        let provider = Provider::new();

        let request = MetadataRequest {
            credential_issuer: ISSUER.to_string(),
            languages: None,
        };
        let response = Endpoint::new(provider).metadata(request).await.expect("response is ok");
        assert_snapshot!("response", response, {
            ".credentials_supported" => insta::sorted_redaction(),
            ".credentials_supported.*.credential_definition.credentialSubject" => insta::sorted_redaction()
        });
    }
}
