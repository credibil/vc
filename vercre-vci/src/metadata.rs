//! # Metadata Endpoint
//!
//! The Credential Issuer Metadata contains information on the Credential Issuer's
//! technical capabilities, supported Credentials, and (internationalized) display
//! information.
//!
//! The Credential Issuer's configuration can be retrieved using the Credential Issuer
//! Identifier.
//!
//! Credential Issuers publishing metadata MUST make a JSON document available at the
//! path formed by concatenating the string `/.well-known/openid-credential-issuer` to
//! the Credential Issuer Identifier. If the Credential Issuer value contains a path
//! component, any terminating / MUST be removed before appending
//! `/.well-known/openid-credential-issuer`.
//!
//! The language(s) in HTTP Accept-Language and Content-Language Headers MUST use the values defined in [RFC3066](https://www.rfc-editor.org/rfc/rfc3066).
//!
//! Below is a non-normative example of a Credential Issuer Metadata request:
//!
//! ```http
//! GET /.well-known/openid-credential-issuer HTTP/1.1
//!     Host: server.example.com
//!     Accept-Language: fr-ch, fr;q=0.9, en;q=0.8, de;q=0.7, *;q=0.5
//! ```

use tracing::{instrument, trace};
pub use vercre_core::metadata as types;
#[allow(clippy::module_name_repetitions)]
pub use vercre_core::vci::{MetadataRequest, MetadataResponse};
use vercre_core::{Callback, Client, Holder, Issuer, Result, Server, Signer, StateManager};

use super::Endpoint;

impl<P> Endpoint<P>
where
    P: Client + Issuer + Server + Holder + StateManager + Signer + Callback + Clone,
{
    /// Metadata request handler.
    ///
    /// # Errors
    ///
    /// Returns an `OpenID4VP` error if the request is invalid or if the provider is
    /// not available.
    pub async fn metadata(&self, request: &MetadataRequest) -> Result<MetadataResponse> {
        let ctx = Context {
            _p: std::marker::PhantomData,
        };
        vercre_core::Endpoint::handle_request(self, request, ctx).await
    }
}

#[derive(Debug)]
struct Context<P> {
    _p: std::marker::PhantomData<P>,
}

impl<P> vercre_core::Context for Context<P>
where
    P: Client + Issuer + Server + Holder + StateManager + Signer,
{
    type Provider = P;
    type Request = MetadataRequest;
    type Response = MetadataResponse;

    // TODO: get callback_id from state
    fn callback_id(&self) -> Option<String> {
        None
    }

    #[instrument]
    async fn process(
        &self, provider: &Self::Provider, request: &Self::Request,
    ) -> Result<Self::Response> {
        trace!("Context::process");

        // TODO: add languages to request
        let credential_issuer = Issuer::metadata(provider, &request.credential_issuer).await?;
        Ok(MetadataResponse { credential_issuer })
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot as assert_snapshot;
    use test_utils::vci_provider::{Provider, ISSUER};

    use super::*;

    #[tokio::test]
    async fn metadata_ok() {
        test_utils::init_tracer();

        let provider = Provider::new();

        let request = MetadataRequest {
            credential_issuer: ISSUER.to_string(),
            languages: None,
        };
        let response = Endpoint::new(provider).metadata(&request).await.expect("response is ok");
        assert_snapshot!("response", response, {
            ".credential_configurations_supported" => insta::sorted_redaction(),
            ".credential_configurations_supported.*.credential_definition.credentialSubject" => insta::sorted_redaction()
        });
    }
}
