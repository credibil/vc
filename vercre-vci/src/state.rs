//! State is used by the library to persist request information between steps
//! in the issuance process.

use chrono::{DateTime, Duration, Utc};
use derive_builder::Builder;
use serde::{Deserialize, Serialize};
use vercre_core::error::Err;
use vercre_core::vci::{CredentialRequest, TokenAuthorizationDetail};
use vercre_core::{err, Result};
pub(crate) enum Expire {
    AuthCode,
    Access,
    Nonce,
}

impl Expire {
    pub(crate) fn duration(&self) -> Duration {
        match self {
            Expire::AuthCode => Duration::minutes(5),
            Expire::Access => Duration::minutes(15),
            Expire::Nonce => Duration::minutes(10),
        }
    }
}

/// State is used to persist request information between issuance steps
/// for the Credential Issuer.
#[derive(Builder, Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub(crate) struct State {
    /// The time this state item should expire.
    #[builder(default = "Utc::now() + Expire::Access.duration()")]
    pub(crate) expires_at: DateTime<Utc>,

    /// The URL of the Credential Issuer.
    pub(crate) credential_issuer: String,

    /// The client_id of the Wallet requesting issuance.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[builder(setter(into, strip_option), default)]
    pub(crate) client_id: Option<String>,

    /// Identifiers of credentials offered to/requested by the Wallet.
    #[builder(default)]
    pub(crate) credential_configuration_ids: Vec<String>,

    /// The subject of the credential Holder.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[builder( default)]
    pub(crate) holder_id: Option<String>,

    /// The callback ID for the current request.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[builder(default)]
    pub(crate) callback_id: Option<String>,

    /// Authorization state.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[builder(default)]
    pub(crate) auth: Option<AuthState>,

    /// Token state.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[builder(default)]
    pub(crate) token: Option<TokenState>,

    /// Deferred step state.
    #[serde(skip_serializing_if = "Option::is_none")]
    #[builder(default)]
    pub(crate) deferred: Option<DeferredState>,
}

impl State {
    /// Returns a new [`StateBuilder`], which can be used to build a [State]
    #[must_use]
    pub(crate) fn builder() -> StateBuilder {
        StateBuilder::default()
    }

    /// Serializes this [`State`] object into a byte array.
    pub(crate) fn to_vec(&self) -> Vec<u8> {
        // TODO: return Result instead of panicking
        match serde_json::to_vec(self) {
            Ok(res) => res,
            Err(e) => panic!("Failed to serialize state: {e}"),
        }
    }

    /// Determines whether state has expired or not.
    pub(crate) fn expired(&self) -> bool {
        self.expires_at.signed_duration_since(Utc::now()).num_seconds() < 0
    }
}

impl TryFrom<&[u8]> for State {
    type Error = vercre_core::error::Error;

    fn try_from(value: &[u8]) -> Result<Self> {
        match serde_json::from_slice::<Self>(value) {
            Ok(res) => {
                if res.expired() {
                    err!(Err::InvalidRequest, "State has expired");
                }
                Ok(res)
            }
            Err(e) => err!(Err::ServerError(e.into()), "Failed to deserialize state"),
        }
    }
}

impl TryFrom<Vec<u8>> for State {
    type Error = vercre_core::error::Error;

    fn try_from(value: Vec<u8>) -> Result<Self> {
        State::try_from(value.as_slice())
    }
}

/// `AuthState` is used to store authorization state.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub(crate) struct AuthState {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(crate) redirect_uri: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub(crate) code_challenge: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub(crate) code_challenge_method: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub(crate) scope: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub(crate) user_code: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub(crate) authorization_details: Option<Vec<TokenAuthorizationDetail>>,
}

/// `TokenState` is used to store token state.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub(crate) struct TokenState {
    /// The access token.
    pub(crate) access_token: String,

    /// The type of token issued (should be "Bearer")
    pub(crate) token_type: String,

    /// The refresh token, issued
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(crate) refresh_token: Option<String>,

    /// The nonce to be used by the Wallet when creating a proof of possession of
    /// the key proof.
    pub(crate) c_nonce: String,

    /// Number denoting the lifetime in seconds of the c_nonce.
    pub(crate) c_nonce_expires_at: DateTime<Utc>,
}

impl TokenState {
    pub(crate) fn c_nonce_expires_in(&self) -> i64 {
        self.c_nonce_expires_at.signed_duration_since(Utc::now()).num_seconds()
    }

    pub(crate) fn c_nonce_expired(&self) -> bool {
        self.c_nonce_expires_in() < 0
    }
}

/// `TokenState` is used to store token state.
#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq, Eq)]
pub(crate) struct DeferredState {
    /// Used to identify a Deferred Issuance transaction. Is used as the
    /// state persistence key.
    pub(crate) transaction_id: String,

    /// Save the CredentialRequest when issuance is deferred.
    pub(crate) credential_request: CredentialRequest,
}

impl AuthState {
    /// Returns a new [`AuthStateBuilder`], which can be used to build a [`State`]
    #[must_use]
    pub(crate) fn builder() -> AuthStateBuilder {
        AuthStateBuilder::new()
    }
}

/// [`AuthStateBuilder`] is used to build a [`AuthState`]
#[derive(Clone, Default)]
pub(crate) struct AuthStateBuilder {
    pub(crate) state: AuthState,
}

impl AuthStateBuilder {
    /// Returns a new [`AuthStateBuilder`]
    #[must_use]
    pub(crate) fn new() -> Self {
        AuthStateBuilder::default()
    }

    /// The client's redirection endpoint if `redirect_uri` was included in the
    /// authorization request. Only used when `grant_type` is "`authorization_code`".
    ///
    /// REQUIRED if the `redirect_uri` parameter was included in the authorization
    /// request and values MUST be identical.
    #[must_use]
    pub(crate) fn redirect_uri(mut self, redirect_uri: String) -> Self {
        self.state.redirect_uri = Some(redirect_uri);
        self
    }

    /// Sets the `code_challenge` properties
    #[must_use]
    pub(crate) fn code_challenge(mut self, challenge: String, method: String) -> Self {
        self.state.code_challenge = Some(challenge);
        self.state.code_challenge_method = Some(method);
        self
    }

    /// Sets the `scope` property
    #[must_use]
    pub(crate) fn scope(mut self, scope: Option<String>) -> Self {
        self.state.scope = scope;
        self
    }

    /// Sets the `user_code` property
    #[must_use]
    pub(crate) fn user_code(mut self, user_code: Option<String>) -> Self {
        self.state.user_code = user_code;
        self
    }

    /// Turns this builder into a [`AuthState`]
    pub(crate) fn build(self) -> AuthState {
        self.state
    }
}

impl TokenState {
    /// Returns a new [`TokenStateBuilder`], which can be used to build a [`TokenState`]
    #[must_use]
    pub(crate) fn builder() -> TokenStateBuilder {
        TokenStateBuilder::new()
    }
}

/// [`TokenStateBuilder`] is used to build a [`TokenState`]
#[derive(Clone, Default)]
pub(crate) struct TokenStateBuilder {
    state: TokenState,
}

impl TokenStateBuilder {
    /// Returns a new [`TokenStateBuilder`]
    #[must_use]
    pub(crate) fn new() -> Self {
        Self {
            state: TokenState {
                token_type: "Bearer".to_string(),
                c_nonce_expires_at: Utc::now() + Expire::Nonce.duration(),
                ..Default::default()
            },
        }
    }

    /// Sets the `access_token` property
    #[must_use]
    pub(crate) fn access_token(mut self, access_token: String) -> Self {
        self.state.access_token = access_token;
        self
    }

    /// Sets the `nonce` property
    #[must_use]
    pub(crate) fn c_nonce(mut self, c_nonce: String) -> Self {
        self.state.c_nonce = c_nonce;
        self
    }

    /// Turns this builder into a [`TokenState`]
    pub(crate) fn build(self) -> TokenState {
        self.state
    }
}
