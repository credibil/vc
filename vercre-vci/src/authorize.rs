//! # Authorization Endpoint
//!
//! The Authorization Endpoint is used in the same manner as defined in [RFC6749](https://www.rfc-editor.org/rfc/rfc6749.html).
//!
//! An Authorization Request is used to request to grant access to the Credential
//! Endpoint.
//!
//! There are two possible ways to request issuance of a specific Credential type in an
//! Authorization Request:
//!
//! 1. Use of the `authorization_details` parameter as defined in [RFC9396](https://www.rfc-editor.org/rfc/rfc9396):
//!
//! ```json
//! [
//!    {
//!       "type": "openid_credential",
//!       "format": "jwt_vc_json",
//!       "credential_definition": {
//!          "type": [
//!             "VerifiableCredential",
//!             "UniversityDegreeCredential"
//!          ]
//!       }
//!    }
//! ]
//! ```
//!
//! 2. Use OAuth 2.0 `scope` parameter:
//!
//! ```http
//! GET /authorize?
//!   response_type=code
//!   &scope=UniversityDegreeCredential
//!   &resource=https://credential-issuer.example.com
//!   &client_id=s6BhdRkqt3
//!   &code_challenge=E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM
//!   &code_challenge_method=S256
//!   &redirect_uri=https%3A%2F%2Fclient.example.org%2Fcb
//! Host: https://server.example.com
//! ```

// LATER: implement `SlowDown` checks/errors

use std::fmt::Debug;
use std::vec;

use chrono::Utc;
use tracing::{instrument, trace};
use vercre_core::error::Err;
use vercre_core::metadata::Issuer as IssuerMetadata;
pub use vercre_core::vci::{
    AuthorizationDetail, AuthorizationRequest, AuthorizationResponse, TokenAuthorizationDetail,
};
use vercre_core::{
    err, gen, Callback, Client, Holder, Issuer, Result, Server, Signer, StateManager,
};

use super::Endpoint;
use crate::state::{AuthState, Expire, State};

impl<P> Endpoint<P>
where
    P: Client + Issuer + Server + Holder + StateManager + Signer + Callback + Clone + Debug,
{
    /// Authorization request handler.
    ///
    /// # Errors
    ///
    /// Returns an `OpenID4VP` error if the request is invalid or if the provider is
    /// not available.
    pub async fn authorize(&self, request: &AuthorizationRequest) -> Result<AuthorizationResponse> {
        //let request = request.into();

        // attempt to get callback_id from state, if pre-auth flow
        let callback_id = if let Some(state_key) = &request.issuer_state {
            let buf = StateManager::get(&self.provider, state_key).await?;
            let state = State::try_from(buf.as_slice())?;
            state.callback_id
        } else {
            None
        };

        let issuer_meta = Issuer::metadata(&self.provider, &request.credential_issuer).await?;
        let cfg_ids = credential_configuration_ids(request, &issuer_meta)?;

        let ctx = Context {
            callback_id,
            credential_configuration_ids: cfg_ids,
        };

        self.handle_request(request, ctx).await
    }
}

#[derive(Debug)]
struct Context {
    callback_id: Option<String>,
    credential_configuration_ids: Vec<String>,
}

impl super::Context for Context {
    type Request = AuthorizationRequest;
    type Response = AuthorizationResponse;

    fn callback_id(&self) -> Option<String> {
        self.callback_id.clone()
    }

    #[instrument]
    async fn verify<P>(&self, provider: &P, request: &Self::Request) -> Result<&Self>
    where
        P: Client + Issuer + Server + Holder + StateManager + Debug,
    {
        trace!("Context::verify");

        let Ok(client_meta) = Client::metadata(provider, &request.client_id).await else {
            err!(Err::InvalidClient, "invalid client_id");
        };
        let server_meta = Server::metadata(provider, &request.credential_issuer).await?;
        let issuer_meta = Issuer::metadata(provider, &request.credential_issuer).await?;

        // 'authorization_code' grant_type allowed (client and server)?
        let client_grant_types = client_meta.grant_types.unwrap_or_default();
        if !client_grant_types.contains(&"authorization_code".to_string()) {
            err!(Err::InvalidRequest, "authorization_code grant not supported for client");
        }
        let server_grant_types = server_meta.grant_types_supported.unwrap_or_default();
        if !server_grant_types.contains(&"authorization_code".to_string()) {
            err!(Err::InvalidRequest, "authorization_code grant not supported for server");
        }

        // holder authorized?
        if request.holder_id.is_empty() {
            err!(Err::AuthorizationPending, "missing holder subject");
        }

        // has a credential been requested?
        if request.authorization_details.is_none() && request.scope.is_none() {
            err!(Err::InvalidRequest, "no credentials requested");
        }

        // verify authorization_details
        'verify_details: for auth_det in request.authorization_details.as_ref().unwrap_or(&vec![]) {
            // we only support `openid_credential` authorization detail requests
            if auth_det.type_ != "openid_credential" {
                err!(Err::InvalidRequest, "invalid authorization_details type");
            }

            let cfg_id_opt = &auth_det.credential_configuration_id;
            let format_opt = &auth_det.format;

            // verify that only one of `credential_configuration_id` or `format` is specified
            if cfg_id_opt.is_some() && format_opt.is_some() {
                err!(
                    Err::InvalidRequest,
                    "`credential_configuration_id` and `format` cannot both be set"
                );
            }
            if cfg_id_opt.is_none() && format_opt.is_none() {
                err!(Err::InvalidRequest, "`credential_configuration_id` or `format` must be set");
            }

            // EITHER: verify requested `credential_configuration_id` is supported
            if let Some(cfg_id) = cfg_id_opt {
                if issuer_meta.credential_configurations_supported.get(cfg_id).is_none() {
                    err!(Err::InvalidRequest, "unsupported credential_configuration_id");
                }
                continue 'verify_details;
            }

            // OR: verify requested `format` and `type` are supported
            if let Some(format) = format_opt {
                let Some(auth_def) = auth_det.credential_definition.as_ref() else {
                    err!(Err::InvalidRequest, "no `credential_definition` specified")
                };

                // check all supported `credential_configurations`
                for cred_cfg in issuer_meta.credential_configurations_supported.values() {
                    if &cred_cfg.format == format
                        && cred_cfg.credential_definition.type_ == auth_def.type_
                    {
                        continue 'verify_details;
                    }
                }

                // couldn't find a matching credential_configuration
                err!(Err::InvalidRequest, "unsupported credential `format` or `type`");
            }
        }

        // verify scope items
        if let Some(scope) = &request.scope {
            'verify_scope: for item in scope.split_whitespace().collect::<Vec<&str>>() {
                for cred_cfg in issuer_meta.credential_configurations_supported.values() {
                    if cred_cfg.scope == Some(item.to_string()) {
                        continue 'verify_scope;
                    }
                }
                err!(Err::InvalidRequest, "scope item {item} is unsupported");
            }
        }

        // TODO: implement `Holder::authorize`
        // - for each requested credential_configuration_id, check if the holder is authorized
        // - save auth_det/scope item for each authorized credential_configuration_id

        // if Holder::authorize(provider, &request.holder_id, &self.credential_configuration_ids).await.is_err() {
        //     err!(Err::AuthorizationPending, "holder is not authorized");
        // }

        // redirect_uri
        let Some(redirect_uri) = &request.redirect_uri else {
            err!(Err::InvalidRequest, "no redirect_uri specified");
        };
        let Some(redirect_uris) = client_meta.redirect_uris else {
            err!(Err::InvalidRequest, "no redirect_uris specified for client");
        };
        if !redirect_uris.contains(redirect_uri) {
            err!(Err::InvalidRequest, "request redirect_uri is not registered");
        }

        // response_type
        if !client_meta.response_types.unwrap_or_default().contains(&request.response_type) {
            err!(Err::UnsupportedResponseType, "the response_type not supported by client");
        }
        if !server_meta.response_types_supported.contains(&request.response_type) {
            err!(Err::UnsupportedResponseType, "response_type not supported by server");
        }

        // code_challenge
        // N.B. while optional in the spec, we require it
        let challenge_methods = server_meta.code_challenge_methods_supported.unwrap_or_default();
        if !challenge_methods.contains(&request.code_challenge_method) {
            err!(Err::InvalidRequest, "unsupported code_challenge_method");
        }
        if request.code_challenge.len() < 43 || request.code_challenge.len() > 128 {
            err!(Err::InvalidRequest, "code_challenge must be between 43 and 128 characters");
        }

        Ok(self)
    }

    #[instrument]
    async fn process<P>(&self, provider: &P, request: &Self::Request) -> Result<Self::Response>
    where
        P: Client + Issuer + Server + Holder + StateManager + Debug,
    {
        trace!("Context::process");

        // save authorization state
        let mut state = State {
            expires_at: Utc::now() + Expire::AuthCode.duration(),
            credential_issuer: request.credential_issuer.clone(),
            client_id: Some(request.client_id.clone()),
            credential_configuration_ids: self.credential_configuration_ids.clone(),
            holder_id: Some(request.holder_id.clone()),
            ..Default::default()
        };

        let mut auth_state = AuthState {
            redirect_uri: request.redirect_uri.clone(),
            code_challenge: Some(request.code_challenge.clone()),
            code_challenge_method: Some(request.code_challenge_method.clone()),
            scope: request.scope.clone(),
            ..Default::default()
        };

        // add `authorization_details` into state
        if let Some(auth_dets) = &request.authorization_details {
            let mut tkn_auth_dets = vec![];

            for auth_det in auth_dets {
                let tkn_auth_det = TokenAuthorizationDetail {
                    authorization_detail: auth_det.clone(),
                    credential_identifiers: None,
                };
                tkn_auth_dets.push(tkn_auth_det.clone());
            }
            auth_state.authorization_details = Some(tkn_auth_dets);
        }

        state.auth = Some(auth_state);

        let code = gen::auth_code();
        StateManager::put(provider, &code, state.to_vec(), state.expires_at).await?;

        // remove offer state
        if let Some(issuer_state) = &request.issuer_state {
            StateManager::purge(provider, issuer_state).await?;
        }

        Ok(AuthorizationResponse {
            code,
            state: request.state.clone(),
            redirect_uri: request.redirect_uri.clone().unwrap_or_default(),
        })
    }
}

#[instrument]
fn credential_configuration_ids(
    request: &AuthorizationRequest, issuer_meta: &IssuerMetadata,
) -> Result<Vec<String>> {
    trace!("credential_configuration_ids");

    let mut cfg_ids = auth_cfg_ids(request, issuer_meta)?;
    cfg_ids.extend(scope_cfg_ids(request, issuer_meta)?);

    Ok(cfg_ids)
}

#[instrument]
fn auth_cfg_ids(
    request: &AuthorizationRequest, issuer_meta: &IssuerMetadata,
) -> Result<Vec<String>> {
    trace!("auth_cfg_ids");

    let Some(authorization_details) = &request.authorization_details else {
        return Ok(vec![]);
    };

    let mut cfg_ids = vec![];

    for auth_det in authorization_details {
        if let Some(cfg_id) = auth_det.credential_configuration_id.clone() {
            // check if requested credential_configuration_id is supported
            if issuer_meta.credential_configurations_supported.get(&cfg_id).is_some() {
                cfg_ids.push(cfg_id);
            }
        } else if let Some(format) = &auth_det.format {
            // check if requested is supported
            for (cfg_id, cred_cfg) in &issuer_meta.credential_configurations_supported {
                // credential_definition must be present
                if &cred_cfg.format == format {
                    let cfg_def = cred_cfg.credential_definition.clone();
                    let auth_def = auth_det.credential_definition.clone().unwrap_or_default();

                    if cfg_def.type_.unwrap_or_default() == auth_def.type_.unwrap_or_default() {
                        cfg_ids.push(cfg_id.clone());
                    }
                }
            }
        }
    }

    Ok(cfg_ids)
}

#[instrument]
fn scope_cfg_ids(
    request: &AuthorizationRequest, issuer_meta: &IssuerMetadata,
) -> Result<Vec<String>> {
    trace!("scope_identifiers");

    let Some(scope) = &request.scope else {
        return Ok(vec![]);
    };
    let mut cfg_ids = vec![];

    for item in scope.split_whitespace().collect::<Vec<&str>>() {
        for (id, cred) in &issuer_meta.credential_configurations_supported {
            if cred.scope == Some(item.to_string()) {
                cfg_ids.push(id.to_owned());
            }
        }
    }

    Ok(cfg_ids)
}

#[cfg(test)]
mod tests {
    use base64ct::{Base64UrlUnpadded, Encoding};
    use insta::assert_yaml_snapshot as assert_snapshot;
    use serde_json::json;
    use sha2::{Digest, Sha256};
    use test_utils::vci_provider::{Provider, ISSUER, NORMAL_USER};
    use test_utils::wallet;

    use super::*;

    #[tokio::test]
    async fn authzn_details() {
        test_utils::init_tracer();

        let provider = Provider::new();

        let auth_dets = json!([{
            "type": "openid_credential",
            "format": "jwt_vc_json",
            "credential_definition": {
                "context": [
                    "https://www.w3.org/2018/credentials/v1",
                    "https://www.w3.org/2018/credentials/examples/v1"
                ],
                "type": [
                    "VerifiableCredential",
                    "EmployeeIDCredential"
                ],
                "credential_subject": {}
            }
        }])
        .to_string();

        let verifier_hash = Sha256::digest("ABCDEF12345");

        // create request
        let body = json!({
            "response_type": "code",
            "client_id": wallet::did(),
            "redirect_uri": "http://localhost:3000/callback",
            "state": "1234",
            "code_challenge": Base64UrlUnpadded::encode_string(&verifier_hash),
            "code_challenge_method": "S256",
            "authorization_details": auth_dets,
            "holder_id": NORMAL_USER,
            "wallet_issuer": ISSUER,
            "callback_id": "1234"
        });
        let mut request =
            serde_json::from_value::<AuthorizationRequest>(body).expect("should deserialize");
        request.credential_issuer = ISSUER.to_string();

        let response =
            Endpoint::new(provider.clone()).authorize(&request).await.expect("response is ok");

        assert_snapshot!("authzn-ok", response, {
            ".code" => "[code]",
        });

        // compare response with saved state
        let buf = StateManager::get(&provider, &response.code).await.expect("state exists");
        let state = State::try_from(buf).expect("state is valid");
        assert_snapshot!("authzn-state", state, {
            ".expires_at" => "[expires_at]",
            ".auth.code" => "[code]",
            ".auth.code_challenge" => "[code_challenge]",
        });
    }

    #[tokio::test]
    async fn scope() {
        test_utils::init_tracer();

        let provider = Provider::new();
        let verifier_hash = Sha256::digest("ABCDEF12345");

        // create request
        let body = json!({
            "response_type": "code",
            "client_id": wallet::did(),
            "redirect_uri": "http://localhost:3000/callback",
            "state": "1234",
            "code_challenge": Base64UrlUnpadded::encode_string(&verifier_hash),
            "code_challenge_method": "S256",
            "scope": "EmployeeIDCredential",
            "holder_id": NORMAL_USER,
            "wallet_issuer": ISSUER,
            "callback_id": "1234"
        });
        let mut request =
            serde_json::from_value::<AuthorizationRequest>(body).expect("should deserialize");
        request.credential_issuer = ISSUER.to_string();

        let response =
            Endpoint::new(provider.clone()).authorize(&request).await.expect("response is ok");
        assert_snapshot!("scope-ok", response, {
            ".code" => "[code]",
        });

        // compare response with saved state
        let buf = StateManager::get(&provider, &response.code).await.expect("state exists");
        let state = State::try_from(buf).expect("state is valid");
        assert_snapshot!("scope-state", state, {
            ".expires_at" => "[expires_at]",
            ".auth.code" => "[code]",
            ".auth.code_challenge" => "[code_challenge]",
        });
    }
}
