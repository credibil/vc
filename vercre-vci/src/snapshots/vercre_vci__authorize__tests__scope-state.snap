---
source: vercre-vci/src/authorize.rs
assertion_line: 470
expression: state
---
expires_at: "[expires_at]"
credential_issuer: "http://credibil.io"
client_id: "did:jwk:eyJjcnYiOiJYMjU1MTkiLCJrdHkiOiJPS1AiLCJ1c2UiOiJlbmMiLCJ4IjoiM0xnOXl2aUFtVERDdVZPeUxYSTNscTlTMnBIbTczeXIzd3dBa2p3Q0FodyJ9"
credential_configuration_ids:
  - EmployeeID_JWT
holder_id: normal_user
auth:
  redirect_uri: "http://localhost:3000/callback"
  code_challenge: "[code_challenge]"
  code_challenge_method: S256
  scope: EmployeeIDCredential
